crates/core/src/metadata_store/providers/etcd.rs

```diff
@@ -106,69 +94,148 @@ impl EtcdMetadataStore {
+        let version_key = Self::version_key(&key);
+
+        let when = match version {
+            None => {
+                vec![Compare::version(version_key.clone(), CompareOp::Equal, 0)]
+            }
+            Some(version) => {
+                vec![Compare::value(
+                    version_key.clone(),
+                    CompareOp::Equal,
+                    version.to_vec(),
+                )]
+            }
+        };
+
+        let txn = Txn::new().when(when).and_then(vec![
+            TxnOp::delete(key, None),
+            TxnOp::delete(version_key, None),
+        ]);
 
         let response = client.txn(txn).await?;
 
         Ok(response.succeeded())
     }
 
+    async fn put_force(
+        &self,
+        client: &mut KvClient,
+        key: Bytes,
+        value: VersionedValue,
+    ) -> Result<(), WriteError> {
+        let version_key = Self::version_key(&key);
+
+        let txn = Txn::new().and_then(vec![
+            TxnOp::put(key, value.value, None),
+            TxnOp::put(version_key, value.version.to_vec(), None),
+        ]);
+
+        client.txn(txn).await?;
+
+        Ok(())
+    }
+
+    async fn delete_force(&self, client: &mut KvClient, key: Bytes) -> Result<(), WriteError> {
+        let version_key = Self::version_key(&key);
+
+        let txn = Txn::new().and_then(vec![
+            TxnOp::delete(key, None),
+            TxnOp::delete(version_key, None),
+        ]);
+
+        client.txn(txn).await?;
+
+        Ok(())
+    }
+
     /// puts a key/value if and only if the current value in store
     /// has the exact given version
     async fn put_if_version_matches(
         &self,
         client: &mut KvClient,
         key: Bytes,
-        value: Bytes,
-        version: i64,
+        value: VersionedValue,
+        version: Option<Version>,
     ) -> Result<bool, WriteError> {
-        let txn = Txn::new()
-            .when(vec![Compare::version(
-                key.clone(),
-                CompareOp::Equal,
-                version,
-            )])
-            .and_then(vec![TxnOp::put(key, value, None)]);
+        let version_key = Self::version_key(&key);
+
+        let when = match version {
+            None => {
+                vec![Compare::version(version_key.clone(), CompareOp::Equal, 0)]
+            }
+            Some(version) => {
+                vec![Compare::value(
+                    version_key.clone(),
+                    CompareOp::Equal,
+                    version.to_vec(),
+                )]
+            }
+        };
+
+        let txn = Txn::new().when(when).and_then(vec![
+            TxnOp::put(key, value.value, None),
+            TxnOp::put(version_key, value.version.to_vec(), None),
+        ]);
 
         let response = client.txn(txn).await?;
 
         Ok(response.succeeded())
     }
+
+    fn version_key(key: &Bytes) -> Bytes {
+        let mut version_key = BytesMut::from_iter(key);
+        version_key.put(&b"::version"[..]);
```

Comment:
Preallocating the `BytesMut` might save us an reallocation when adding the `::version` suffix.